### ISSUE TITLE
Go API

### ISSUE THREAD

**Issue by alonsovidales** [tensorflow/tensorflow#10]
Hi!,
I'm really interested on contribute in my spare time on this project, it  would be great for me port the Python libraries to Go, or help with this task, the problem is that after read the "Contributing guidelines" I still don't know how to do it.
I tried to sing the Individual CLA, but I get an error that says: "You must be an owner of the contributors group in order to submit this CLA.".

Btw, are you working, or do you plan to implement soon the Go libs?

Some related projects that I implemented in Go: https://github.com/alonsovidales/pit - This is a distributed recommender system based in the "Adaptive Bootstrapping of Recommender Systems Using Decision Trees" paper by Yahoo. The site of the project: http://pitia.info/
And the Go ML: Some machine learning algorithms in Go: https://github.com/alonsovidales/go_ml I also implemented Neural Networks with CUDA: https://github.com/alonsovidales/go_ml/tree/cuda_implementation

Thanks,

**Comment by jamesliu96** [comment]
:+1:

**Comment by pcasaretto** [comment]
I'd be interested in helping @alonsovidales !

**Comment by jamesliu96** [comment]
Idk if I'm qualified but I'd like to do whatever to help @alonsovidales !

**Comment by mrkaspa** [comment]
I can help too because I need this library in Go

**Comment by chai2010** [comment]
+1/0

**Comment by alonsovidales** [comment]
I still have no answer, but we could start working on the libs, and if Google wants, they can use the code, if not we can release the libs as a separate project :) I think that this shouldn't be a problem since the project is licensed using an Apache 2.0 license.
My main concern is that we could be doing the same that they are doing the same in parallel, so, by the moment we can keep reading and trying to understand how the currently available libs works, and if this weekend we don't have any news, we can prepare, plan and distribute the different tasks to do the libraries.

What do you think?
I'm going to send an e-mail to the golang-nuts mail list.

**Comment by jamesliu96** [comment]
agree! wait till weekend then we can decide.

**Comment by hachi8833** [comment]
:+1:

**Comment by 1l0** [comment]
Great move!

**Comment by vincentvanhoucke** [comment]
The current state of Go is that we have a few people who've expressed interest internally in contributing language bindings, but nothing concrete at the moment. I would suggest that if you got started on this, which would be awesome, you could whip up a proposal (preferably in the form of prototype code) and run the design by the discuss mailing list.

**Comment by alonsovidales** [comment]
Thanks @vincentvanhoucke , then, let's try to prepare a prototype and let's see how it works :) I had been digging a bit in the Python libs, and it shouldn't be too difficult to prepare the Go libraries.

Btw, which is the discuss mailing list?, I can't find it.

I love the project and it would be awesome for me be able to contribute.

**Comment by alonsovidales** [comment]
Found the discuss mailing list, sorry :)

**Comment by davidzchen** [comment]
Should we keep this issue open to track this?

**Comment by alonsovidales** [comment]
Sure, better keep this issue open so other people can find it easily.

I think that it would be better to move the conversation about the prototype, etc to another place, we can:
- Use the official "discuss mailing list"
- Create a new mailing list just for this
- Use the issue tracker of a GitHub repo, I created this repo just to start playing around: https://github.com/alonsovidales/tensorflow

Regarding to the prototype, my proposal is to use one of the examples:
http://tensorflow.org/tutorials/mnist/beginners/index.md
And studying how the Python libs works, port just all the involved code to execute the example, since an example like the one for beginners covers most of the functionality, and can give us a great overview of how to port all the libraries.
To reproduce the behaviour of the Python libs, the best option would be to use TDD, we can go method by method and based on the input/output of the python code prepare the test on Go.

We can distribute the tasks to prepare the prototype or work in parallel and put all our ideas together at the end. Since for me the goal here is try to get a general idea about how to implement the final version, tasks that we have to do, etc, I think that work in parallel can be a good approach.

What do you think?

**Comment by babakgh** [comment]
:+1:

**Comment by alonsovidales** [comment]
Just a little update of what I have been researching this weekend. I couldn't find too much time for this, but I made some small progresses.

I have been researching how the Python libs works internally, and basically, the logic to generate the graph and plan how to execute it, etc is managed from Python that sends all the graph to be executed to the C++ libraries, this libs take the graph and executes all the nodes of the graph. They are working on a pure C++ version of this logic, but it is still not ready.
In order to execute the C++ code from python they use SWIG: www.swig.org and to build the system Bazel: http://bazel.io/ .
The good news is that SWIG also supports Go ( http://www.swig.org/Doc2.0/Go.html ) and Bazel introduces support for Go ( http://bazel.io/docs/be/go.html )  after the 0.1.1 version.
The bad news are that I couldn't build the system using Bazel 0.1.1 by many different problems with the linker, etc that I'm trying to resolve. And the other problem is that we have to create the .i files for SWIG in order to prepare the build for Go, I almost have this part done, but I'm also having a lot of problems.
The main issue is that I have no experience with Bazel and SWIG, so I'm spending most of the time reading documentation and trying to fix stupid problems caused because of I misunderstand something from the docu.

In case of I finally can't prepare a wrapper using SWIG, I think that I'm going to try to prepare the prototype using cgo directly, I already have a small proof of concept that generates the session in Go, but it would be much better if the libs works with SWIG and Bazel as the Python libs does.

I'm not having too much spare time lately, but it is being super interesting for me to research how all this system is done, the code is pretty well documented, and easy to understand.

**Comment by nathangoulding** [comment]
It seems like Bazel doesn't yet have C/C++ interop for Go so cgo may be the only option unless Bazel adds support for it soon.
http://bazel.io/docs/be/go.html

**Comment by alonsovidales** [comment]
I think that it would be possible to do a small workaround using a sh_binary to compile run swig like they are doing with the Python libs:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/BUILD#L21
But by the moment for the PoC I'm using an ugly MakeFile:
https://github.com/alonsovidales/tensorflow/blob/master/tensorflow/go/makefile
And setting the flags for the linker here:
https://github.com/alonsovidales/tensorflow/blob/master/tensorflow/go/cgoflags.go
And also a wrapper to access the C++ libraries from C:
https://github.com/alonsovidales/tensorflow/tree/master/tensorflow/go/wrapper
But this is just for the prototype, I plan to move all to Bazel + SWIG :)

**Comment by nathangoulding** [comment]
I'm still not 100% clear on the long-term delineation between the external library's responsibilities and the underlying TF C++ API.

From http://www.tensorflow.org/api_docs/cc/index.html, this sentence implies that maybe at some point in the future the C++ libraries will add the support to build the graphs: "TensorFlow's public C++ API includes only the API for executing graphs, as of version 0.5."

It seems like the supported way outside of python is to still use python to build the graph, then use tf.train.write_graph() to write the graph to a file, and then load it into a session in your language of choice and then run it using the C++ API.

Are you planning to port the graph building mechanism into Go? It seems like it might be a good idea to get the consensus of the TF dev team before embarking on something as ambitious as that if they're planning on supporting this in the public API at some point in the future.

**Comment by alonsovidales** [comment]
Well, my idea is to build the graph in the Go libs as it is being builded by the Python libs. They are working on the C++ code, but by the moment it is a WIP and it is, by the moment, unstable.

I'm not sure about their future plans, but from my point of view have this process in Python is not a bad approach since the tensors takes a lot of advantage of the dynamic typing and operators overloading that Python offers, this makes the code more simple to write and to understand, this is being one of the biggest problem I'm having porting the code to Go. I also think that keep the graph building process in Python is a good idea since this process requires almost no time compared with the operations that are implemented in C++, so scarify simplicity in favour of performance don't makes too much sense. I think that they make the good choice here.
I'm planning to port the process to Go by two reasons:
 - You could create a pure Go application to train your model, that could be interesting for some situations.
 - I'm working on just a PoC, and I prefer to explore all the code in order to get a global idea of how all the different pieces works together.
And well, port the process is going to be funny for me :)

**Comment by vladimirvivien** [comment]
@alonsovidales any movement on your PoC ?

**Comment by alonsovidales** [comment]
No :'( I had a personal problem and I couldn't spend a single min on personal projects during last month, I'll try to resume it on next week.

**Comment by tmc** [comment]
I have a start of a swig-based binding started here: [tmc/tensorflow/go_bindings](https://github.com/tmc/tensorflow/tree/go_bindings) -- mostly cargo-culted and incomplete.

**Comment by MiCHiLU** [comment]
:+1:

**Comment by Lerenn** [comment]
:+1:

**Comment by dave-andersen** [comment]
tmc sent a pull request - https://github.com/tensorflow/tensorflow/pull/1237 - which we should be able to get in with a few iterations.

**Comment by alonsovidales** [comment]
I finally have the C++ example of the Image Recognition tutorial ported to Go. Implement all the Graph construction, constants and other stuff was relatively easy since there is available the list of operations perfectly defined here:
 - https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/ops.pbtxt

Here you have the Go port of the tuto example:
https://github.com/alonsovidales/tensorflow/blob/go_bindings_tensors/tensorflow/examples/label_image_go/main.go
that corresponds with:
https://github.com/alonsovidales/tensorflow/blob/go_bindings_tensors/tensorflow/examples/label_image/main.cc

This is how looks like running a picture of Ceres my Labrador:
```
avidales@bigbrain:~/go-src/src/github.com/tensorflow/tensorflow/tensorflow/examples/label_image_go$ go run main.go data/ceres_paris.jpg
Labrador retriever : 0.9383322
American Staffordshire terrier : 0.009385344
Rhodesian ridgeback : 0.007571198
Chesapeake Bay retriever : 0.0027833031
golden retriever : 0.0026763931
```

I'm going to clean the code and include a section for Go libs on the tutorial.

**Comment by girving** [comment]
@dave-andersen What's the status of this?

**Comment by dave-andersen** [comment]
Closed by #1771.

**Comment by tmc** [comment]
#1771 appears to be in the 'go' branch but not master -- what's the plan to merge that in?

**Comment by dave-andersen** [comment]
Realistically, TBD.  I've maybe found someone @ google who might be interested in taking ownership of the bigger picture of TF + go.  I think we're of a mind that the current state isn't what we need yet -- it's not integrated with the normal tests, nor is it go-buildable in the normal way.  Those are the major blockers to getting it into the main branch, but I think there's a further concern that because of the amount of magic that still happens only in Python with shapes & gradients, we're not at a point where we can have the Go bindings we'd like as a first-class supported language.  There's some ongoing work looking at improving our ability to support more languages, and I think some of the momentum on Go is stalled waiting on that.

**Comment by tmc** [comment]
Fair enough. I personally just wanted a way to execute graphs vs have python-level support.

**Comment by girving** [comment]
Let's leave this bug open then, to go along with the bunch of other open language bugs that are much less far along.

**Comment by Mistobaan** [comment]
I tried the go branch after a rebase and it works on my mac. Now I would like to propose some changes with a PR (https://github.com/Mistobaan/tensorflow/tree/go). But clearly the current go branch is behind respect to master and the PR is huge.
Should we rebase the main go branch or merge it in master and keep continuing the development there? 
What is the best feasible strategy to keep patches coming into the go interface?

**Comment by alonsovidales** [comment]
Hi @Mistobaan , I've been a bit out the last months because I'm moving to Dublin. @dave-andersen mentions on a previous comment that someone at Google could be interested on taking ownership of the Go development, the code was lent in the Go branch waiting for the integration with Jenkins and for other people to play with it and provide some feedback, I'm not sure about what is the current status.
I took a look at your commit and it seems really interesting, I think that you could make a PR to the current branch and we  can care to rebase it in a future :)
I'm using the Go libs on some personal projects and I got feedback from other people that are using them, It would be awesome for us to count with improvements.

**Comment by Mistobaan** [comment]
@alonsovidales see #3258

**Comment by damienstanton** [comment]
Hi @Mistobaan and @alonsovidales, for my own project I have been working on merging the master updates into the go branch [here](https://github.com/damienstanton/tensorflow/tree/go).

TF build is fine, however `go test` is currently failing:
```
go git:go✎ go test
# github.com/tensorflow/tensorflow/tensorflow/contrib/go
/var/folders/r3/ndy82hd54c5b2yxqys6mx88w0000gn/T/go-build408123681/github.com/tensorflow/tensorflow/tensorflow/contrib/go/_obj/tensorflow_wrap.cxx:1
850:52: error: elaborated type refers to a typedef
./../../../tensorflow/core/public/tensor_c_api.h:89:3: note: declared here
FAIL    github.com/tensorflow/tensorflow/tensorflow/contrib/go [build failed]
```
I am new to working with swig. Any ideas here?

**Comment by damienstanton** [comment]
Follow-up: I've since resolved that issue (an enum that was improperly `typdef`'d). I am working out one other type assertion bug, but otherwise looking good. I'll update once all tests pass on my branch

**Comment by damienstanton** [comment]
Okay - good to go with all updates from `master` in `go`. [PR Here](https://github.com/tensorflow/tensorflow/pull/3419).

**Comment by damienstanton** [comment]
@Mistobaan see new PR for the same ☝️

**Comment by damienstanton** [comment]
FYI #3542 resolves a build error I experienced on the Go branch.

**Comment by jmhodges** [comment]
I've just published a PR to bump the go branch to the latest master and do some minor fixups in #4223

**Comment by jmhodges** [comment]
I'm not sure what's left to do in the go branch to get it merged to master. I might be able to help!

**Comment by jmhodges** [comment]
Updated #4223.

I think there was a goal of getting this to build with `go get` and `go install`. Currently, the generated protobuf doesn't have the right import paths to build correctly. (It's a version of the problem in https://github.com/golang/protobuf/issues/230)

We're going to have to probably do some sed work to fix it.

**Comment by jhseu** [comment]
Before you do more work on that branch, keep in mind that we have in-progress Go bindings that the TensorFlow team is making :)

**Comment by jmhodges** [comment]
That's cool! Didn't know! Maybe missed a comment. Anything I can do to help?

**Comment by pyanfield** [comment]
@jhseu , do you have any plan about when we can get the official release?

**Comment by jhseu** [comment]
We've got a handle on them :) For obvious reasons, we're particular about the structure of the Go API and some particulars (like autosubmitting generated protobuf code).

We need some additions to the C API (work-in-progress) to complete them. Should be in about a week or two.

**Comment by jhseu** [comment]
Note that it's here:
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/go

We're making a more reasonable import path for people to use.

**Comment by jmhodges** [comment]
Cool! What's the plan for the protobuf generated packages? I'm currently just copying the proto files into my own build to pop out the Go code I want and they just barely don't fit there trivially (because protoc-gen-proto has some rules about where its generated code goes and what import paths they refer to each other with)

**Comment by jmhodges** [comment]
Oh, man, after weeks of hunting, I just found https://github.com/golang/protobuf/issues/139 so now I understand how it'll work

**Comment by asimshankar** [comment]
@jmhodges : So far whatever functionality the [Go API provides](https://godoc.org/github.com/tensorflow/tensorflow/tensorflow/go)  doesn't require protobuf. But of course, the functionality is limited at this point.

If y'all have any feedback on the API as it is today and what you'd find most useful in terms of additional functionality or changes, we'd greatly appreciate that.

**Comment by asimshankar** [comment]
Hello folks, the Go API is in a reasonably usable shape at this point. Graphs can be imported, executed, constructed (using primitives or functions generated from the TensorFlow op definitions etc.).

The API is by no means _complete_, but at this point I'm tempted to close out this generic "Go API" issue with the expectation that as the Go API sees real usage, new issues will be filed with specific feature requests and bug reports.

**Comment by helinwang** [comment]
I created this package to enable tensorboard for tensorflow golang api. Commenting here in case anyone interested https://github.com/helinwang/tfsum

**Comment by damienstanton** [comment]
@helinwang, very cool! If you need any help with this I would be happy to contribute

**Comment by helinwang** [comment]
save and load tensorflow tensors using tensorflow golang api: https://gist.github.com/helinwang/7782c6b2815c334c77653fc0e52b6069

**Comment by sdeoras** [comment]
Hi Folks, I am just curious why wrapper util functions in "op" folder don't seem to take node name as input argument. This prevents me from calling functions twice in a scope. For instance, I can't define two constants in a graph. There are no errors when I explicitly define constants or placeholders using tf.OpSpec by providing values for the Name field.

**Comment by asimshankar** [comment]
@sdeoras : The alternative for now is to use [`Subscope`](https://godoc.org/github.com/tensorflow/tensorflow/tensorflow/go/op#Scope.SubScope), which is [used in the label image example](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/go/example_inception_inference_test.go#L211) as well (resulting in constants named `make_batch/Const`, `size/Const` etc. in the graph).

The answer to your question though is just that it felt less cumbersome to require a `name` argument on every op addition when it can often be elided. We'd want it to be an "optional" argument. One option is to do what we do in C++ graph construction API and add a [`WithOpName`](https://github.com/tensorflow/tensorflow/blob/r0.12/tensorflow/cc/framework/scope.h#L104) to the `Scope` type.

If you have strong opinions or suggestions, please do share (though perhaps in a separate issue focused on this aspect). In the mean time, hopefully the alternative suggested above lets you make progress.

**Comment by sdeoras** [comment]
@asimshankar : Thanks for your quick reply. Since it is easier for me to think in terms of graphs and operations on them, I am currently just rewriting some of the wrapper functions with node name as additional argument to avoid getting errors from underlying C++ code. I do, however, feel the same that providing names for every op declaration is cumbersome and hope that future versions of TF will no longer make this a requirement.

**Comment by sdeoras** [comment]
Hi Folks, I am not sure if this thread is the best place to ask TF GO API related questions. If not, please suggest a link to the right forum.

So I can't seem to get following code to work. It complains on second use of op.Reshape at runtime but allows two calls to op.Const... what am I doing wrong?

`	scope := op.NewScope()
	x := op.Placeholder(scope, tf.Double)
	s := op.Const(scope, []int32{3, 3})
	y := op.Reshape(scope, x, s)
	yInv := op.MatrixInverse(scope, y)
	s2 := op.Const(scope.SubScope("scope.a"), []int32{9})
	yInvReshaped := op.Reshape(scope.SubScope("scope.b"), yInv, s2)

	graph, err := scope.Finalize()
	if err != nil {
		t.Fatal(err)
	}
`

**Comment by asimshankar** [comment]
Please use stackoverflow for general questions, or file new github issues for bugs/feature requests. In this case I've filed #6799 - I think I know what the problem is, will update there.

**Comment by sdeoras** [comment]
Hi Folks, I am running into another issue with shape argument in op.VarHandleOp. Pl. advise if this should be filed as a bug.

```

package bugs

import (
	tf "github.com/tensorflow/tensorflow/tensorflow/go"
	"github.com/tensorflow/tensorflow/tensorflow/go/op"
	"testing"
)

func TestVarHandleOpShape(t *testing.T) {
	scope := op.NewScope()
	_ = op.VarHandleOp(scope, tf.Int32, []int64{2, 3})
	_, err := scope.Finalize()
	if err != nil {
		t.Fatal(err)
	}
}

/*
=== RUN   TestVarHandleOpShape
--- FAIL: TestVarHandleOpShape (0.03s)
        bugs_test.go:14: failed to add operation "VarHandleOp": AttrValue had value with type 'list(int)' when 'shape' expected
                         for attr 'shape'
                        ; NodeDef: VarHandleOp = VarHandleOp[_class=[], container="", dtype=DT_INT32, shape=[2, 3], shared_name=""](); Op<name=VarHandleOp; signature= -> resource:resource; attr=container:string,default=""; attr=shared_name:string,default=""; attr=dtype:type; attr=shape:shape; is_stateful=true> (Stacktrace: goroutine 5 [running]:
                runtime/debug.Stack(0x0, 0x0, 0x0)
                        /usr/lib/golang/src/runtime/debug/stack.go:24 +0x79
                github.com/tensorflow/tensorflow/tensorflow/go/op.(*Scope).UpdateErr(0xc42000c420, 0x53e4eb, 0xb, 0x7cce00, 0xc420054040)
                        /home/sdeoras/go/src/github.com/tensorflow/tensorflow/tensorflow/go/op/scope.go:113 +0x72
                github.com/tensorflow/tensorflow/tensorflow/go/op.(*Scope).AddOperation(0xc42000c420, 0x53e4eb, 0xb, 0x53e4eb, 0xb, 0x0, 0x0, 0x0, 0xc42000c450, 0xc42000c3f0)
                        /home/sdeoras/go/src/github.com/tensorflow/tensorflow/tensorflow/go/op/scope.go:78 +0xf9
                github.com/tensorflow/tensorflow/tensorflow/go/op.VarHandleOp(0xc42000c420, 0xc400000003, 0xc4200106f0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x171e4d44, 0xc420051f78)
                        /home/sdeoras/go/src/github.com/tensorflow/tensorflow/tensorflow/go/op/wrappers.go:14045 +0x27d
                bitbucket.hgst.com/x/tensorflow.git/tfutil.TestVarHandleOpShape(0xc4200b8180)
                        /home/sdeoras/go/src/bitbucket.hgst.com/x/tensorflow.git/tfutil/bugs_test.go:11 +0x96
                testing.tRunner(0xc4200b8180, 0x54f7f0)
                        /usr/lib/golang/src/testing/testing.go:610 +0x81
                created by testing.(*T).Run
                        /usr/lib/golang/src/testing/testing.go:646 +0x2ec
                )
FAIL
exit status 1
FAIL    bitbucket.hgst.com/x/tensorflow.git/tfutil      0.130s
 */

```

**Comment by asimshankar** [comment]
Yeah, shape attributes handling [wasn't finished](https://github.com/tensorflow/tensorflow/blob/8f6cb22/tensorflow/go/graph.go#L263), but should be easy. Feel free to file a new issue for this (and don't hesitate to file new issues for other things instead of adding to this closed one).

**Comment by sdeoras** [comment]
@helinwang Great work with enabling TensorBoard! Could you please include a license file for us to be able to use it for commercial purposes?

**Comment by helinwang** [comment]
@sdeoras Glad to know that the library helps! Added 3-clause BSD.

**Comment by sdeoras** [comment]
Thank you @helinwang

**Comment by 4d55397500** [comment]
Apologies in advance for not following this issue thoroughly, but what's the state of the Go binding? If it's finished can someone point me to tutorial implementations, for example linear regression over the MNIST dataset? Thanks

**Comment by lipixun** [comment]
@4d55397500 I believe the go binding of tensorflow is still under heavy developing. See: https://godoc.org/github.com/tensorflow/tensorflow/tensorflow/go

It is not stable enough and lots of features haven't been implemented.

(I've just compiled the latest version and sadly speaking that it can not pass the builtin test on mac os x. A signal `kill` occurred when running tests.)

**Comment by asimshankar** [comment]
@lipixun @4d55397500 : The Go APIs available today are intended to be usable and tests for both Linux and OS X are included in the continuous builds on Jenkins. 

@4d55397500 : The APIs are a little low level though. They should be sufficient to build things like networks for MNIST and linear regression models on top of, but at this time there are no plans for writing convenience higher level APIs.
